### ISSUE TITLE
ObservationEventSampler not working

### ISSUE THREAD

**Issue by HealthyPear** [gammapy/gammapy#5542]
**Gammapy version**

1.2

**Bug description**

Use of ObservationEventSampler fails because it's attempting to get the maximum value from the class, not from its data

**Expected behavior**

According to the source code it should be compatible with observations providing RadMax2D (otherwise there would be no reason to look for a maximum value)

**To Reproduce**

```python
from gammapy.data import ObservationsEventsSampler
observation = # any pointlike observation from a data store
model = # any sky model
obs_evt_sampler.simulate_observation(observation, models=Models(models=model))
```

```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[14], line 1
----> 1 obs_evt_sampler.simulate_observation(observations[0], models=Models(models=model))

File [~/Software/gammapy/gammapy/gammapy/data/simulate.py:69](http://localhost:8888/lab/tree/~/Software/gammapy/gammapy/gammapy/data/simulate.py#line=68), in ObservationsEventsSampler.simulate_observation(self, observation, models)
     64 from gammapy.datasets import ObservationEventSampler
     66 sampler = ObservationEventSampler(
     67     **self.sampler_kwargs, dataset_kwargs=self.dataset_kwargs
     68 )
---> 69 observation = sampler.run(observation, models=models)
     71 if self.outdir is not None:
     72     observation.write(
     73         self.outdir [/](http://localhost:8888/) f"obs_{observation.obs_id}.fits",
     74         overwrite=self.overwrite,
     75     )

File [~/Software/gammapy/gammapy/gammapy/datasets/simulate.py:681](http://localhost:8888/lab/tree/~/Software/gammapy/gammapy/gammapy/datasets/simulate.py#line=680), in ObservationEventSampler.run(self, observation, models, dataset_name)
    657 def run(self, observation, models=None, dataset_name=None):
    658     """Sample events for given observation and signal models.
    659 
    660     The signal distribution is sampled from the given models
   (...)
    678         A copy of the input observation with event list filled.
    679     """
--> 681     dataset = create_map_dataset_from_observation(
    682         observation, models, dataset_name, **self.dataset_kwargs
    683     )
    684     events = super().run(dataset, observation)
    686     observation = deepcopy(observation)

File [~/Software/gammapy/gammapy/gammapy/datasets/map.py:228](http://localhost:8888/lab/tree/~/Software/gammapy/gammapy/gammapy/datasets/map.py#line=227), in create_map_dataset_from_observation(observation, models, dataset_name, energy_axis_true, energy_axis, energy_bin_per_decade_max, spatial_width, spatial_width_max, spatial_bin_size, spatial_bin_size_min)
    225 from gammapy.makers import MapDatasetMaker
    227 if spatial_width is None:
--> 228     spatial_width = _default_width(observation, spatial_width_max)
    229 if spatial_bin_size is None:
    230     spatial_bin_size = _default_binsz(observation, spatial_bin_size_min)

File [~/Software/gammapy/gammapy/gammapy/datasets/map.py:173](http://localhost:8888/lab/tree/~/Software/gammapy/gammapy/gammapy/datasets/map.py#line=172), in _default_width(observation, spatial_width_max)
    170 def _default_width(observation, spatial_width_max=12 * u.deg):
    171     # width estimated from the rad_max or the offset_max
    172     if observation.rad_max is not None:
--> 173         width = 2.0 * np.max(observation.rad_max)
    174     else:
    175         width = 2.0 * observation.psf.axes["offset"].edges[-1]

TypeError: unsupported operand type(s) for *: 'float' and 'RadMax2D'
```

**Comment by HealthyPear** [comment]
when attempting to fix the issue it cascades to `_default_binsz` as it expects a psf IRF, but in the case of a point source, there is no PSF, but the observations I am using are from a point-like observation so the PSF is not expected to be in the DL3 products, only rad max

**Comment by HealthyPear** [comment]
I found also another issue: `dataset_kwargs` is set to None by default, but later expanded like a mapping in `create_map_dataset_from_observation` so it will of course crash unless one defines an empty mapping

`self.dataset_kwargs = dataset_kwargs if dataset_kwargs else {}`

**Comment by HealthyPear** [comment]
Said this,

obs_evt_sampler = ObservationsEventsSampler()
obs_evt_sampler.simulate_observation(obs, models=Models(models=model))

is currently failing from main (with those 2 little fixes on top) for any observation, even from HESS and CTA example data

**Comment by HealthyPear** [comment]
The additional issue that prevents this class to work even on full-enclosure-based observations is the way true energy is defined to build the energy axis for the mapdataset

`etrue = observation.psf.axes[0].edges` is too long when getting the energy resolution should be `center`

like this the following example works,

```python
from gammapy.data import Observation, ObservationsEventsSampler
from gammapy.modeling.models import Models

obs_evt_sampler = ObservationsEventsSampler(outdir=None)

model = SkyModel(spectral_model=spectral_model, spatial_model=spatial_model, name="target_source")

# from a real observation
data_store = DataStore.from_dir("$GAMMAPY_DATA/cta-1dc/index/gps")
real_observation = data_store.obs(int(observations.ids[0]))

# sky model of a Crab-like source at a convenient celestial coordinate
spectral_model = LogParabolaSpectralModel(
    amplitude=3.39e-11 * u.Unit("cm-2 s-1 TeV-1"),
    alpha=2.51,
    beta=0.21,
    reference=1 * u.TeV,
)

spatial_model = PointSpatialModel(
    lon_0=real_observation.pointing.get_icrs().ra,
    lat_0=real_observation.pointing.get_icrs().dec,
    frame="icrs"
)

fake_obs = obs_evt_sampler.simulate_observation(real_observation, models=Models(models=model))

# from an originally fake observation with some dummy (full-enclosure) IRFs
irfs = {"aeff": some_observation.aeff, "edisp": some_observation.edisp, "psf":some_observation.psf, "bkg":some_observation.bkg}

fake_observation =  Observation.create(
    pointing=some_pointing,
    livetime=some_time,
    irfs=irfs,
    location=a_location,
)

fake_obs = obs_evt_sampler.simulate_observation(fake_obs, models=Models(models=model))
```

in both cases `test_observation.events` will contain the sampled events.

The code above might be a new tutorial for the use of this class from full-enclosure IRFs, but the limitation for point-like IRFs still stands.

An alternative might be to simulate from full-enclosure and then analyze as point-like, but I am not sure how to do it making sure that the final estimated statistics is correct

**Comment by HealthyPear** [comment]
I'll open a PR to fix the use for full-enclosure and an issue regarding the point-like use case.

**Comment by QRemy** [comment]
Thanks @HealthyPear for looking closely into this.

> `etrue = observation.psf.axes[0].edges` is too long when getting the energy resolution should be `center`

This is solved in #5533. But if you can make a bug fix PR before it would be good.

**Comment by HealthyPear** [comment]
> This is solved in https://github.com/gammapy/gammapy/pull/5533. But if you can make a bug fix PR before it would be good.

I guess it might make sense as this fix is a very small one and I guess that PR targets a broader issue?
